File: generate_diagram.py
#!/usr/bin/env python3
"""
Generate a visual diagram of the React Visibility Service Architecture
"""

import matplotlib.pyplot as plt
import matplotlib.patches as patches
from matplotlib.patches import FancyBboxPatch, ConnectionPatch
import numpy as np

# Set up the figure with high DPI for better quality
plt.rcParams['figure.dpi'] = 300
plt.rcParams['savefig.dpi'] = 300
fig, ax = plt.subplots(1, 1, figsize=(16, 12))
ax.set_xlim(0, 16)
ax.set_ylim(0, 12)
ax.axis('off')

# Colors
colors = {
    'react': '#61dafb',
    'visibility': '#ff6b6b', 
    'state_machine': '#4ecdc4',
    'core_observer': '#45b7d1',
    'api': '#f9ca24',
    'text': '#2c3e50',
    'border': '#34495e'
}

# Title
ax.text(8, 11.5, 'React Visibility Service Architecture', 
        fontsize=24, fontweight='bold', ha='center', color=colors['text'])

# React Components Section
ax.text(2, 10.5, 'React Components', fontsize=16, fontweight='bold', 
        color=colors['text'], ha='center')

# Component boxes
component_box = FancyBboxPatch((0.5, 9.5), 3, 0.8, 
                              boxstyle="round,pad=0.1", 
                              facecolor=colors['react'], 
                              edgecolor=colors['border'], linewidth=2)
ax.add_patch(component_box)
ax.text(2, 9.9, 'Component', fontsize=12, ha='center', fontweight='bold')

# Hook boxes
hook_box = FancyBboxPatch((0.5, 8.5), 3, 0.8, 
                          boxstyle="round,pad=0.1", 
                          facecolor=colors['react'], 
                          edgecolor=colors['border'], linewidth=2)
ax.add_patch(hook_box)
ax.text(2, 8.9, 'useVisibility Hook', fontsize=12, ha='center', fontweight='bold')

handler_box = FancyBboxPatch((0.5, 7.5), 3, 0.8, 
                            boxstyle="round,pad=0.1", 
                            facecolor=colors['react'], 
                            edgecolor=colors['border'], linewidth=2)
ax.add_patch(handler_box)
ax.text(2, 7.9, 'VisibilityHandler', fontsize=12, ha='center', fontweight='bold')

watch_box = FancyBboxPatch((0.5, 6.5), 3, 0.8, 
                          boxstyle="round,pad=0.1", 
                          facecolor=colors['react'], 
                          edgecolor=colors['border'], linewidth=2)
ax.add_patch(watch_box)
ax.text(2, 6.9, 'useWatch Hook', fontsize=12, ha='center', fontweight='bold')

# Visibility Instance Section
ax.text(8, 10.5, 'Visibility Instance', fontsize=16, fontweight='bold', 
        color=colors['text'], ha='center')

instance_box = FancyBboxPatch((6.5, 9.5), 3, 0.8, 
                             boxstyle="round,pad=0.1", 
                             facecolor=colors['visibility'], 
                             edgecolor=colors['border'], linewidth=2)
ax.add_patch(instance_box)
ax.text(8, 9.9, 'VisibilityInstance', fontsize=12, ha='center', fontweight='bold')

# State Machine
state_box = FancyBboxPatch((6.5, 8.5), 3, 0.8, 
                           boxstyle="round,pad=0.1", 
                           facecolor=colors['state_machine'], 
                           edgecolor=colors['border'], linewidth=2)
ax.add_patch(state_box)
ax.text(8, 8.9, 'State Machine', fontsize=12, ha='center', fontweight='bold')

# States
open_state = FancyBboxPatch((5.5, 7.5), 2, 0.6, 
                            boxstyle="round,pad=0.05", 
                            facecolor=colors['state_machine'], 
                            edgecolor=colors['border'], linewidth=1)
ax.add_patch(open_state)
ax.text(6.5, 7.8, 'State: open', fontsize=10, ha='center')

close_state = FancyBboxPatch((8.5, 7.5), 2, 0.6, 
                             boxstyle="round,pad=0.05", 
                             facecolor=colors['state_machine'], 
                             edgecolor=colors['border'], linewidth=1)
ax.add_patch(close_state)
ax.text(9.5, 7.8, 'State: close', fontsize=10, ha='center')

# API Methods
api_box = FancyBboxPatch((6.5, 6.5), 3, 0.8, 
                         boxstyle="round,pad=0.1", 
                         facecolor=colors['api'], 
                         edgecolor=colors['border'], linewidth=2)
ax.add_patch(api_box)
ax.text(8, 6.9, 'API Methods', fontsize=12, ha='center', fontweight='bold')

# API method details
open_method = FancyBboxPatch((5.5, 5.5), 1.8, 0.6, 
                             boxstyle="round,pad=0.05", 
                             facecolor=colors['api'], 
                             edgecolor=colors['border'], linewidth=1)
ax.add_patch(open_method)
ax.text(6.4, 5.8, 'open()', fontsize=10, ha='center')

close_method = FancyBboxPatch((7.2, 5.5), 1.8, 0.6, 
                              boxstyle="round,pad=0.05", 
                              facecolor=colors['api'], 
                              edgecolor=colors['border'], linewidth=1)
ax.add_patch(close_method)
ax.text(8.1, 5.8, 'close()', fontsize=10, ha='center')

reset_method = FancyBboxPatch((8.9, 5.5), 1.8, 0.6, 
                              boxstyle="round,pad=0.05", 
                              facecolor=colors['api'], 
                              edgecolor=colors['border'], linewidth=1)
ax.add_patch(reset_method)
ax.text(9.8, 5.8, 'reset()', fontsize=10, ha='center')

# State Machine Engine Section
ax.text(13, 10.5, 'State Machine Engine', fontsize=16, fontweight='bold', 
        color=colors['text'], ha='center')

engine_box = FancyBboxPatch((11.5, 9.5), 3, 0.8, 
                            boxstyle="round,pad=0.1", 
                            facecolor=colors['state_machine'], 
                            edgecolor=colors['border'], linewidth=2)
ax.add_patch(engine_box)
ax.text(13, 9.9, '@scoped-observer/react-state-machine', fontsize=10, ha='center', fontweight='bold')

create_box = FancyBboxPatch((11.5, 8.5), 3, 0.8, 
                            boxstyle="round,pad=0.1", 
                            facecolor=colors['state_machine'], 
                            edgecolor=colors['border'], linewidth=2)
ax.add_patch(create_box)
ax.text(13, 8.9, 'createMachine', fontsize=12, ha='center', fontweight='bold')

transitions_box = FancyBboxPatch((11.5, 7.5), 3, 0.8, 
                                boxstyle="round,pad=0.1", 
                                facecolor=colors['state_machine'], 
                                edgecolor=colors['border'], linewidth=2)
ax.add_patch(transitions_box)
ax.text(13, 7.9, 'State Transitions', fontsize=12, ha='center', fontweight='bold')

# Core Observer Section
ax.text(8, 4.5, 'Core Observer System', fontsize=16, fontweight='bold', 
        color=colors['text'], ha='center')

core_box = FancyBboxPatch((6.5, 3.5), 3, 0.8, 
                          boxstyle="round,pad=0.1", 
                          facecolor=colors['core_observer'], 
                          edgecolor=colors['border'], linewidth=2)
ax.add_patch(core_box)
ax.text(8, 3.9, '@scoped-observer/core', fontsize=12, ha='center', fontweight='bold')

observer_box = FancyBboxPatch((5.5, 2.5), 2, 0.6, 
                              boxstyle="round,pad=0.05", 
                              facecolor=colors['core_observer'], 
                              edgecolor=colors['border'], linewidth=1)
ax.add_patch(observer_box)
ax.text(6.5, 2.8, 'ScopedObserver', fontsize=10, ha='center')

entity_box = FancyBboxPatch((8.5, 2.5), 2, 0.6, 
                            boxstyle="round,pad=0.05", 
                            facecolor=colors['core_observer'], 
                            edgecolor=colors['border'], linewidth=1)
ax.add_patch(entity_box)
ax.text(9.5, 2.8, 'EventEntity', fontsize=10, ha='center')

events_box = FancyBboxPatch((6.5, 1.5), 3, 0.8, 
                            boxstyle="round,pad=0.1", 
                            facecolor=colors['core_observer'], 
                            edgecolor=colors['border'], linewidth=2)
ax.add_patch(events_box)
ax.text(8, 1.9, 'Event Management', fontsize=12, ha='center', fontweight='bold')

# Arrows showing data flow
# Component to Visibility Instance
arrow1 = ConnectionPatch((3.5, 9.9), (6.5, 9.9), "data", "data",
                        arrowstyle="->", shrinkA=5, shrinkB=5, 
                        mutation_scale=20, fc=colors['border'], lw=2)
ax.add_patch(arrow1)

# Visibility Instance to State Machine
arrow2 = ConnectionPatch((8, 8.5), (8, 8.5), "data", "data",
                        arrowstyle="->", shrinkA=5, shrinkB=5, 
                        mutation_scale=20, fc=colors['border'], lw=2)
ax.add_patch(arrow2)

# State Machine to Engine
arrow3 = ConnectionPatch((9.5, 8.9), (11.5, 8.9), "data", "data",
                        arrowstyle="->", shrinkA=5, shrinkB=5, 
                        mutation_scale=20, fc=colors['border'], lw=2)
ax.add_patch(arrow3)

# Visibility Instance to Core Observer
arrow4 = ConnectionPatch((8, 6.5), (8, 3.5), "data", "data",
                        arrowstyle="->", shrinkA=5, shrinkB=5, 
                        mutation_scale=20, fc=colors['border'], lw=2)
ax.add_patch(arrow4)

# State transitions
# open to close
arrow5 = ConnectionPatch((6.5, 7.2), (8.5, 7.2), "data", "data",
                        arrowstyle="->", shrinkA=5, shrinkB=5, 
                        mutation_scale=15, fc=colors['border'], lw=1.5)
ax.add_patch(arrow5)
ax.text(7.5, 7.0, 'ON_CLOSE', fontsize=9, ha='center', color=colors['text'])

# close to open
arrow6 = ConnectionPatch((8.5, 7.8), (6.5, 7.8), "data", "data",
                        arrowstyle="->", shrinkA=5, shrinkB=5, 
                        mutation_scale=15, fc=colors['border'], lw=1.5)
ax.add_patch(arrow6)
ax.text(7.5, 8.0, 'ON_OPEN', fontsize=9, ha='center', color=colors['text'])

# API connections
arrow7 = ConnectionPatch((6.4, 5.5), (6.5, 7.2), "data", "data",
                        arrowstyle="->", shrinkA=5, shrinkB=5, 
                        mutation_scale=15, fc=colors['border'], lw=1.5)
ax.add_patch(arrow7)

arrow8 = ConnectionPatch((8.1, 5.5), (8.5, 7.2), "data", "data",
                        arrowstyle="->", shrinkA=5, shrinkB=5, 
                        mutation_scale=15, fc=colors['border'], lw=1.5)
ax.add_patch(arrow8)

arrow9 = ConnectionPatch((9.8, 5.5), (8.5, 7.2), "data", "data",
                        arrowstyle="->", shrinkA=5, shrinkB=5, 
                        mutation_scale=15, fc=colors['border'], lw=1.5)
ax.add_patch(arrow9)

# Add legend
legend_elements = [
    patches.Patch(color=colors['react'], label='React Components'),
    patches.Patch(color=colors['visibility'], label='Visibility Instance'),
    patches.Patch(color=colors['state_machine'], label='State Machine'),
    patches.Patch(color=colors['core_observer'], label='Core Observer'),
    patches.Patch(color=colors['api'], label='API Methods')
]

ax.legend(handles=legend_elements, loc='upper right', bbox_to_anchor=(0.98, 0.98))

# Add data flow description
ax.text(0.5, 0.8, 'Data Flow: Component → useVisibility → VisibilityInstance → State Machine → Core Observer', 
        fontsize=12, fontweight='bold', color=colors['text'])

ax.text(0.5, 0.5, 'Key Features: Scoped Management, State Persistence, Event System, Type Safety, Memory Management', 
        fontsize=10, color=colors['text'])

plt.tight_layout()
plt.savefig('visibility_service_diagram.png', bbox_inches='tight', dpi=300, facecolor='white')
plt.show()

print("✅ Diagram saved as 'visibility_service_diagram.png'")

File: generate_diagram.js
#!/usr/bin/env node
/**
 * Generate a visual diagram of the React Visibility Service Architecture
 * Using Node.js Canvas for PNG generation
 */

const { createCanvas, registerFont } = require('canvas');
const fs = require('fs');

// Create canvas
const width = 1600;
const height = 1200;
const canvas = createCanvas(width, height);
const ctx = canvas.getContext('2d');

// Colors
const colors = {
    background: '#f8f9fa',
    react: '#61dafb',
    visibility: '#ff6b6b',
    stateMachine: '#4ecdc4',
    coreObserver: '#45b7d1',
    api: '#f9ca24',
    text: '#2c3e50',
    border: '#34495e',
    white: '#ffffff',
    shadow: 'rgba(0,0,0,0.1)'
};

// Helper function to draw rounded rectangle
function drawRoundedRect(x, y, width, height, radius, fillColor, strokeColor, lineWidth = 2) {
    ctx.beginPath();
    ctx.moveTo(x + radius, y);
    ctx.lineTo(x + width - radius, y);
    ctx.quadraticCurveTo(x + width, y, x + width, y + radius);
    ctx.lineTo(x + width, y + height - radius);
    ctx.quadraticCurveTo(x + width, y + height, x + width - radius, y + height);
    ctx.lineTo(x + radius, y + height);
    ctx.quadraticCurveTo(x, y + height, x, y + height - radius);
    ctx.lineTo(x, y + radius);
    ctx.quadraticCurveTo(x, y, x + radius, y);
    ctx.closePath();
    
    if (fillColor) {
        ctx.fillStyle = fillColor;
        ctx.fill();
    }
    
    if (strokeColor) {
        ctx.strokeStyle = strokeColor;
        ctx.lineWidth = lineWidth;
        ctx.stroke();
    }
}

// Helper function to draw text with shadow
function drawText(text, x, y, fontSize, fontFamily, color, align = 'center', baseline = 'middle') {
    ctx.font = `${fontSize}px ${fontFamily}`;
    ctx.fillStyle = color;
    ctx.textAlign = align;
    ctx.textBaseline = baseline;
    
    // Add shadow
    ctx.shadowColor = 'rgba(0,0,0,0.3)';
    ctx.shadowBlur = 4;
    ctx.shadowOffsetX = 2;
    ctx.shadowOffsetY = 2;
    
    ctx.fillText(text, x, y);
    
    // Reset shadow
    ctx.shadowColor = 'transparent';
    ctx.shadowBlur = 0;
    ctx.shadowOffsetX = 0;
    ctx.shadowOffsetY = 0;
}

// Helper function to draw arrow
function drawArrow(fromX, fromY, toX, toY, color, lineWidth = 3) {
    ctx.strokeStyle = color;
    ctx.lineWidth = lineWidth;
    ctx.lineCap = 'round';
    
    // Draw line
    ctx.beginPath();
    ctx.moveTo(fromX, fromY);
    ctx.lineTo(toX, toY);
    ctx.stroke();
    
    // Draw arrowhead
    const angle = Math.atan2(toY - fromY, toX - fromX);
    const arrowLength = 15;
    const arrowAngle = Math.PI / 6;
    
    ctx.beginPath();
    ctx.moveTo(toX, toY);
    ctx.lineTo(
        toX - arrowLength * Math.cos(angle - arrowAngle),
        toY - arrowLength * Math.sin(angle - arrowAngle)
    );
    ctx.moveTo(toX, toY);
    ctx.lineTo(
        toX - arrowLength * Math.cos(angle + arrowAngle),
        toY - arrowLength * Math.sin(angle + arrowAngle)
    );
    ctx.stroke();
}

// Clear canvas
ctx.fillStyle = colors.background;
ctx.fillRect(0, 0, width, height);

// Title
drawText('React Visibility Service Architecture', width/2, 80, 36, 'Arial', colors.text, 'center', 'middle');
drawText('Complete system overview and data flow diagram', width/2, 130, 20, 'Arial', colors.text, 'center', 'middle');

// React Components Section
drawText('⚛️ React Components', 200, 200, 24, 'Arial', colors.text, 'center', 'middle');

// Component boxes
drawRoundedRect(50, 230, 300, 60, 10, colors.white, colors.react);
drawText('Component', 200, 260, 18, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(50, 310, 300, 60, 10, colors.white, colors.react);
drawText('useVisibility Hook', 200, 340, 18, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(50, 390, 300, 60, 10, colors.white, colors.react);
drawText('VisibilityHandler', 200, 420, 18, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(50, 470, 300, 60, 10, colors.white, colors.react);
drawText('useWatch Hook', 200, 500, 18, 'Arial', colors.text, 'center', 'middle');

// Visibility Instance Section
drawText('👁️ Visibility Instance', 800, 200, 24, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(650, 230, 300, 60, 10, colors.white, colors.visibility);
drawText('VisibilityInstance', 800, 260, 18, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(650, 310, 300, 60, 10, colors.white, colors.stateMachine);
drawText('State Machine', 800, 340, 18, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(650, 390, 300, 60, 10, colors.white, colors.api);
drawText('API Methods', 800, 420, 18, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(650, 470, 300, 60, 10, colors.white, colors.stateMachine);
drawText('States: open | close', 800, 500, 18, 'Arial', colors.text, 'center', 'middle');

// State Machine Engine Section
drawText('⚙️ State Machine Engine', 1400, 200, 24, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(1250, 230, 300, 60, 10, colors.white, colors.stateMachine);
drawText('@scoped-observer/react-state-machine', 1400, 260, 16, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(1250, 310, 300, 60, 10, colors.white, colors.stateMachine);
drawText('createMachine', 1400, 340, 18, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(1250, 390, 300, 60, 10, colors.white, colors.stateMachine);
drawText('State Transitions', 1400, 420, 18, 'Arial', colors.text, 'center', 'middle');

// Core Observer Section
drawText('🔗 Core Observer System', 800, 600, 24, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(650, 630, 300, 60, 10, colors.white, colors.coreObserver);
drawText('@scoped-observer/core', 800, 660, 18, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(550, 710, 200, 50, 8, colors.white, colors.coreObserver);
drawText('ScopedObserver', 650, 735, 16, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(950, 710, 200, 50, 8, colors.white, colors.coreObserver);
drawText('EventEntity', 1050, 735, 16, 'Arial', colors.text, 'center', 'middle');

drawRoundedRect(650, 780, 300, 60, 10, colors.white, colors.coreObserver);
drawText('Event Management', 800, 810, 18, 'Arial', colors.text, 'center', 'middle');

// Data Flow Section
drawText('🔄 Data Flow Architecture', width/2, 900, 28, 'Arial', colors.text, 'center', 'middle');

// Flow boxes
const flowY = 950;
const flowSpacing = 200;

drawRoundedRect(100, flowY, 150, 50, 25, colors.react, colors.border);
drawText('Component', 175, flowY + 25, 16, 'Arial', colors.white, 'center', 'middle');

drawRoundedRect(300, flowY, 150, 50, 25, colors.react, colors.border);
drawText('useVisibility', 375, flowY + 25, 16, 'Arial', colors.white, 'center', 'middle');

drawRoundedRect(500, flowY, 150, 50, 25, colors.visibility, colors.border);
drawText('VisibilityInstance', 575, flowY + 25, 16, 'Arial', colors.white, 'center', 'middle');

drawRoundedRect(700, flowY, 150, 50, 25, colors.stateMachine, colors.border);
drawText('State Machine', 775, flowY + 25, 16, 'Arial', colors.white, 'center', 'middle');

drawRoundedRect(900, flowY, 150, 50, 25, colors.coreObserver, colors.border);
drawText('Core Observer', 975, flowY + 25, 16, 'Arial', colors.white, 'center', 'middle');

// Draw arrows between flow boxes
drawArrow(250, flowY + 25, 300, flowY + 25, colors.border);
drawArrow(450, flowY + 25, 500, flowY + 25, colors.border);
drawArrow(650, flowY + 25, 700, flowY + 25, colors.border);
drawArrow(850, flowY + 25, 900, flowY + 25, colors.border);

// Main arrows
drawArrow(350, 260, 650, 260, colors.border, 4);
drawArrow(800, 310, 800, 630, colors.border, 4);
drawArrow(950, 340, 1250, 340, colors.border, 4);

// State transitions
drawArrow(650, 500, 950, 500, colors.border, 2);
drawText('ON_CLOSE', 800, 480, 14, 'Arial', colors.text, 'center', 'middle');

drawArrow(950, 470, 650, 470, colors.border, 2);
drawText('ON_OPEN', 800, 450, 14, 'Arial', colors.text, 'center', 'middle');

// API connections
drawArrow(700, 420, 700, 500, colors.border, 2);
drawArrow(800, 420, 800, 500, colors.border, 2);
drawArrow(900, 420, 900, 500, colors.border, 2);

// Legend
const legendY = 1050;
const legendSpacing = 250;

// React
ctx.fillStyle = colors.react;
ctx.fillRect(50, legendY, 20, 20);
drawText('React Components', 100, legendY + 10, 16, 'Arial', colors.text, 'left', 'middle');

// Visibility
ctx.fillStyle = colors.visibility;
ctx.fillRect(300, legendY, 20, 20);
drawText('Visibility Instance', 350, legendY + 10, 16, 'Arial', colors.text, 'left', 'middle');

// State Machine
ctx.fillStyle = colors.stateMachine;
ctx.fillRect(550, legendY, 20, 20);
drawText('State Machine', 600, legendY + 10, 16, 'Arial', colors.text, 'left', 'middle');

// Core Observer
ctx.fillStyle = colors.coreObserver;
ctx.fillRect(800, legendY, 20, 20);
drawText('Core Observer', 850, legendY + 10, 16, 'Arial', colors.text, 'left', 'middle');

// API
ctx.fillStyle = colors.api;
ctx.fillRect(1050, legendY, 20, 20);
drawText('API Methods', 1100, legendY + 10, 16, 'Arial', colors.text, 'left', 'middle');

// Save the image
const buffer = canvas.toBuffer('image/png');
fs.writeFileSync('visibility_service_diagram.png', buffer);

console.log('✅ Diagram saved as visibility_service_diagram.png');
console.log('📊 Image dimensions:', width, 'x', height);
console.log('🎨 Colors used:', Object.keys(colors).length);